File: org.wxd.boot.jscript/src/test/java/test/MyRandom.java
package test;

import lombok.extern.slf4j.Slf4j;

import java.io.Serializable;

/**
 * 自定义随机数
 *
 * @author: Troy.Chen(無心道, 15388152619)
 * @version: 2022-05-05 11:14
 **/
@Slf4j
public final class MyRandom implements Serializable {

    static final ThreadLocal<MyRandom> LOCAL = ThreadLocal.withInitial(() -> new MyRandom());

    /** 当前线程安全的 */
    public static MyRandom current() {
        return LOCAL.get();
    }

    volatile long oldSeed = 0l;
    volatile long addSeed = 0l;

    private MyRandom() {
        oldSeed = System.currentTimeMillis() ^ 999999999L;
        refreshSeed();
    }

    /**
     * 只能用一次随机数种子
     *
     * @param seed
     */
    public MyRandom setSeed(long seed) {
        oldSeed = seed;
        return this;
    }

    /** 刷新随机数种子 */
    protected void refreshSeed() {
        oldSeed = (oldSeed * 9301 + addSeed++) & 233280L;
    }

    /** 随机范围 */
    public int nextInt(int min, int max) {
        if (max <= min) {
            return min;
        }
        int tmp = max - min;
        float next = next();
        return min + Math.round(tmp * next);
    }

    /** float 下一个 */
    public float nextFloat(float min, float max) {
        if (max <= min) {
            return min;
        }
        float tmp = max - min;
        float next = next();
        return min + tmp * next;
    }

    /** 自定义随机算法, 0 ~ 1 */
    public float next() {
        float v = oldSeed / 233280.0f;
        if (v < 0) {
            v = 0;
        }
        if (v > 1) {
            v = 1;
        }
        refreshSeed();
        return v;
    }

}

File: wxdgaming.boot.jscript/src/test/resources/MyRandom.js
var oldSeed = Date.now() ^ 999999999;
var addSeed = 0;

/** 刷新随机数种子 */
function refreshSeed() {
    oldSeed = (oldSeed * 9301 + addSeed++) & 233280;
}

function nextInt(min, max, seed) {
    if (max <= min) {
        return min;
    }
    var tmp = max - min;
    var next00 = next(seed);
    return Math.round(min + Math.round(tmp * next00));
}

/**
 * 自定义随机算法, 0 ~ 1
 */
function next(seed) {
    if (!(seed === null
        || typeof (seed) === "undefined")) {
        oldSeed = seed;
    } else {
        refreshSeed();
    }

    var v = oldSeed / 233280.0;
    if (v < 0) {
        v = 0;
    }
    if (v > 1) {
        v = 1;
    }
    return v;
}